tsuru/config/config.go

```diff
@@ -37,51 +59,63 @@ func bootstrapConfig() *ConfigType {
 		return newDefaultConf()
 	}
 	if err != nil {
-		fmt.Fprintf(stderr, "Could not read %q: %v\nContinuing without config", configPath, err)
+		fmt.Fprintf(stderr, "Could not read %q: %v\nContinuing without config\n", configPath, err)
 		return nil
 	}
 	defer file.Close()
 
 	rawContent, err := io.ReadAll(file)
 	if err != nil {
-		fmt.Fprintf(stderr, "Could not read %q: %v\nContinuing without config", configPath, err)
+		fmt.Fprintf(stderr, "Could not read %q: %v\nContinuing without config\n", configPath, err)
 		return nil
 	}
 
 	config := ConfigType{}
 	if err := json.Unmarshal(rawContent, &config); err != nil {
-		nowTimeStr := time.Now().UTC().Format("2006-01-02_15:04:05")
+		nowTimeStr := nowUTC().Format("2006-01-02_15:04:05")
 		backupFilePath := configPath + "." + nowTimeStr + ".bak"
 		fmt.Fprintf(stderr, "Error parsing %q: %v\n", configPath, err)
 		fmt.Fprintf(stderr, "Backing up current file to %q. A new configuration will be saved.\n", backupFilePath)
 		filesystem().Rename(configPath, backupFilePath)
 		return newDefaultConf()
 	}
 
-	config.lastChanges = config.LastUpdate
+	config.originalContent = string(rawContent)
+
+	// Convert any older config
+	// ...
+
+	// Mandatory fields
+	if config.ClientSelfUpdater.LatestManifestURL == "" {
+		config.ClientSelfUpdater.LatestManifestURL = defaultLatestManifestURL
+	}
+
 	return &config
 }
 
-func getConfig() *ConfigType {
-	if config == nil {
-		config = bootstrapConfig()
+// GetConfig() returns a *ConfigType singleton.
+func GetConfig() *ConfigType {
+	if privConfig == nil {
+		privConfig = bootstrapConfig()
 	}
-	return config
+
+	return privConfig
 }
 
 func (c *ConfigType) hasChanges() bool {
 	if c == nil {
 		return false
 	}
-	return c.LastUpdate.Before(c.lastChanges)
+	jsonConfig, _ := json.Marshal(c)
```

Comment:
i don't think we should ever ignore error checking, it would be easy enough to make `hasChanges()` return `(bool, error)` and change `SaveChangesNoPrint()` handle the error accordingly

Reply:
Agreed. If this ConfigType is extended in the future with some type prone to errors, it would be destructive without that check.